rfcs/0024-hard-forks.md

```diff
@@ -0,0 +1,172 @@
+
+Upgrades will happen twice per year, on March 1st 00:00:00 UTC and October 1st
+00:00:00 UTC. Clients that do not include support for a scheduled fork will
+**halt all progress** at that slot. This ensures out of date clients don't
+continue on a mostly dead and therefore vulnerable network. Client that do
+support the upgrade will, of course, execute the transition at that time/slot.
+
+Mistakes in hard forks can be very bad, but are not intrinsically more dangerous
+than client upgrades. However, they are more dangerous in that nodes that do not
+upgrade could end up on a dead fork without realizing it. TODO is an on-chain
+signaling mechanism possible so clients can detect this condition?
+
+Because mistakes may be catastrophic and upgrades may involve large changes,
+this process is very conservative. The timeline will be as follows:
+
+1. Beginning of previous era:
+  * Discussion, design and implementation of features for next era begins. This
+    can happen in parallel for different features, and should be mostly done
+    before the finalization process begins.
+1. Six weeks from era start:
+  * A public testnet starts, forked off the mainnet at the current block. At
+    this point it has no changes from the previous era, other than being
+    segregated. This will be used as a dry-run for the upcoming fork. We may
+    need to modify delegations in order to ensure sufficient stake is online.
+  * Feature specification lock. All features proposed to go into this hard fork
+    have been submitted to the RFC process. Everything submitted must have a
+    working, tested implementation.
+2. Five weeks from era start:
+  * All proposals are accepted or rejected/pushed to the next era.
+3. Four weeks from era start:
+  * Testnet era advances and the new features begin the final burn-in test.
+    Builds should be available implementing everything, at a "release candidate"
+    level of quality.
+  * The Coda team and general public monitor the testnet, use it, and try to
+    break it.
+4. Up until one week from era start:
+  * For every issue we discover, if fixing it would require another hard fork we
+    disable the associated feature unless the change is very simple. Disabling a
+    feature requires starting a new era on the testnet. If fixing it is only a
+    code change we're a little more lenient: if the fix is minor we make it and
+    continue, if it's major we still drop the feature.
+5. One week from era start:
+  * Mainnet client builds that support the new era are released and all nodes
+    are expected to upgrade. If no further intervention occurs the transition
+    will proceed as planned.
+  * No further changes are allowed. If a bug that should block era transition is
+    found we delay the era start by another week, releasing new clients, and
+    return to step 4.
+6. Era start:
+  * All changes are active on the mainnet and the burn-in testnet is retired.
+    Pop the confetti.
+
```

Comment:
If this schedule is followed, there should be a checklist somewhere, maybe with an automated mechanism to enforce it.